File: src/visual-graph/builder/find-node-by-id.test.ts
import { describe, expect, test } from "vitest";

import { DIRECTION } from "../direction.js";
import { NODE_KIND } from "../node-kind.js";
import { SUBGRAPH_KIND } from "../subgraph-kind.js";
import { VISUAL_ELEMENT_TYPE } from "../visual-element-type.js";
import type { VisualElement } from "../visual-element.js";
import type { VisualNode } from "../visual-node.js";
import { findNodeById } from "./find-node-by-id.js";

function leafNode(id: string, name = id): VisualNode {
  return {
    type: VISUAL_ELEMENT_TYPE.Node,
    id,
    kind: NODE_KIND.Variable,
    name,
    line: 1,
    endLine: null,
    isJsxElement: false,
    unused: false,
    declarationKind: null,
    initIsFunction: false,
  };
}

const elements: VisualElement[] = [
  leafNode("a"),
  {
    type: VISUAL_ELEMENT_TYPE.Subgraph,
    id: "s1",
    kind: SUBGRAPH_KIND.Function,
    line: 1,
    endLine: null,
    direction: DIRECTION.RL,
    ownerNodeId: "n_owner",
    ownerName: "owner",
    elements: [
      leafNode("b"),
      {
        type: VISUAL_ELEMENT_TYPE.Subgraph,
        id: "s2",
        kind: SUBGRAPH_KIND.If,
        line: 1,
        endLine: null,
        direction: DIRECTION.RL,
        elements: [leafNode("c")],
      },
    ],
  },
  leafNode("d"),
];

describe("findNodeById", () => {
  test.each([
    { id: "a", expected: "a" },
    { id: "b", expected: "b" },
    { id: "c", expected: "c" },
    { id: "d", expected: "d" },
  ])("finds top-level and nested node $id", ({ id, expected }) => {
    expect(findNodeById(elements, id)?.id).toBe(expected);
  });

  test("returns null when id is absent", () => {
    expect(findNodeById(elements, "missing")).toBeNull();
  });

  test("returns null on an empty element list", () => {
    expect(findNodeById([], "a")).toBeNull();
  });

  test("ignores subgraph ids (only matches node ids)", () => {
    expect(findNodeById(elements, "s1")).toBeNull();
    expect(findNodeById(elements, "s2")).toBeNull();
  });
});

File: crates/unsnarl-visual-graph/src/builder/find_node_by_id_test.rs
//! Sibling tests for [`find_node_by_id`].

use super::find_node_by_id;
use crate::direction::Direction;
use crate::visual_element::VisualElement;
use crate::visual_element_type::{NodeTypeTag, SubgraphTypeTag};
use crate::visual_node::{BindingExtras, BindingNodeKind, BindingVisualNode, VisualNode};
use crate::visual_subgraph::{
    ControlExtras, ControlSubgraphKind, ControlVisualSubgraph, OwnedExtras, OwnedSubgraphKind,
    OwnedVisualSubgraph, VisualSubgraph,
};

fn leaf(id: &str) -> VisualElement {
    VisualElement::Node(VisualNode::Binding(BindingVisualNode {
        r#type: NodeTypeTag::Node,
        id: id.to_string(),
        name: id.to_string(),
        line: 1,
        end_line: None,
        is_jsx_element: false,
        unused: false,
        kind: BindingNodeKind::ConstBinding,
        extras: BindingExtras::Variable {
            init_is_function: false,
        },
    }))
}

fn function_subgraph(id: &str, elements: Vec<VisualElement>) -> VisualElement {
    VisualElement::Subgraph(VisualSubgraph::Owned(OwnedVisualSubgraph {
        r#type: SubgraphTypeTag::Subgraph,
        id: id.to_string(),
        kind: OwnedSubgraphKind::Function,
        line: 1,
        end_line: None,
        direction: Direction::RL,
        extras: OwnedExtras::Function {
            owner_node_id: Some("n_owner".to_string()),
            owner_name: "owner".to_string(),
        },
        elements,
    }))
}

fn if_subgraph(id: &str, elements: Vec<VisualElement>) -> VisualElement {
    VisualElement::Subgraph(VisualSubgraph::Control(ControlVisualSubgraph {
        r#type: SubgraphTypeTag::Subgraph,
        id: id.to_string(),
        line: 1,
        end_line: None,
        direction: Direction::RL,
        elements,
        kind: ControlSubgraphKind::If,
        extras: ControlExtras::None {},
    }))
}

fn build_elements() -> Vec<VisualElement> {
    let inner = if_subgraph("s2", vec![leaf("c")]);
    let function = function_subgraph("s1", vec![leaf("b"), inner]);
    vec![leaf("a"), function, leaf("d")]
}

#[test]
fn finds_top_level_node() {
    let mut elements = build_elements();
    let found = find_node_by_id(&mut elements, "a").expect("node found");
    assert_eq!(found.id(), "a");
}

#[test]
fn finds_node_one_level_down() {
    let mut elements = build_elements();
    let found = find_node_by_id(&mut elements, "b").expect("node found");
    assert_eq!(found.id(), "b");
}

#[test]
fn finds_node_two_levels_down() {
    let mut elements = build_elements();
    let found = find_node_by_id(&mut elements, "c").expect("node found");
    assert_eq!(found.id(), "c");
}

#[test]
fn finds_top_level_after_nested_subgraph() {
    let mut elements = build_elements();
    let found = find_node_by_id(&mut elements, "d").expect("node found");
    assert_eq!(found.id(), "d");
}

#[test]
fn returns_none_when_id_is_absent() {
    let mut elements = build_elements();
    assert!(find_node_by_id(&mut elements, "missing").is_none());
}

#[test]
fn returns_none_on_empty_element_list() {
    let mut elements: Vec<VisualElement> = Vec::new();
    assert!(find_node_by_id(&mut elements, "a").is_none());
}

#[test]
fn ignores_subgraph_ids() {
    // `find_node_by_id` only matches nodes, never subgraphs.
    let mut elements = build_elements();
    assert!(find_node_by_id(&mut elements, "s1").is_none());
    assert!(find_node_by_id(&mut elements, "s2").is_none());
}
